biospecimen-tracking/api/src/main/java/org/phenotips/data/internal/controller/BiospecimensController.java

```diff
@@ -0,0 +1,379 @@
+    public void writeJSON(Patient patient, JSONObject jsonObject, Collection<String> selectedFieldNames)
+    {
+        if (selectedFieldNames != null && !BiospecimenData.PROPERTIES.containsAll(selectedFieldNames)) {
+            if (this.logger.isDebugEnabled()) {
+                this.logger.debug("Some of the given selectedFieldNames [{}] do not belong to the allowed values [{}]",
+                    selectedFieldNames, BiospecimenData.PROPERTIES);
+            }
+            return;
+        }
+
+        PatientData<BiospecimenData> biospecimenData = patient.getData(getName());
+
+        if (biospecimenData == null || biospecimenData.size() == 0) {
+            return;
+        }
+
+        Collection<String> propertiesToInclude = propertyNamesToIterateOver(selectedFieldNames);
+
+        DateFormat dateFormat =
+            new SimpleDateFormat(this.configurationManager.getActiveConfiguration().getISODateFormat());
+
+        JSONArray biospecimensJsonArray = new JSONArray();
+
+        for (BiospecimenData biospecimen : biospecimenData) {
+            if (biospecimen == null) {
+                continue;
+            }
+            JSONObject obj = getJSONObjectFromBiospecimenData(biospecimen, propertiesToInclude, dateFormat);
+            if (obj != null) {
+                biospecimensJsonArray.put(obj);
+            }
+        }
+
+        if (biospecimensJsonArray.length() > 0) {
+            jsonObject.put(getName(), biospecimensJsonArray);
+        }
+    }
+
+    @Override
+    public PatientData<BiospecimenData> readJSON(JSONObject jsonObject)
+    {
+        if (jsonObject == null || jsonObject.optJSONArray(getName()) == null) {
+            return null;
+        }
+
+        JSONArray biospecimensJsonArray = jsonObject.getJSONArray(getName());
+
+        List<BiospecimenData> result = new LinkedList<>();
+
+        DateFormat jsonDateFormat =
+            new SimpleDateFormat(this.configurationManager.getActiveConfiguration().getISODateFormat());
+
+        try {
+            for (Object biospecimenObject : biospecimensJsonArray) {
+                BiospecimenData biospecimen = parseBiospecimenObject(biospecimenObject, jsonDateFormat);
+                if (biospecimen != null) {
+                    result.add(biospecimen);
+                }
+            }
+        } catch (ParseException e) {
+            this.logger.error("Unable to parse JSON data [{}]", e.getMessage());
+            return null;
+        }
+
+        return new IndexedPatientData<>(getName(), result);
+    }
+
+    /**
+     * Gets an EntityReference to the xWiki class representing BiospecimenData in the patient document.
+     * @return a EntityReference object
+     */
+    public static EntityReference getClassReference()
+    {
+        return new EntityReference(CLASS_NAME, EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+    }
+
+    /***************************************** Private Methods ****************************************/
```

Comment:
I don't think this is necessary. We usually define public methods first, then private methods. In the future, we will standardize on this.